src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double RESIZE_CONSTANT = 1.5;
+    private static final int EMPTY_SIZE = 0;
+    private Object[] elementData;
+    private int size;
+    private int newCapacity;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        resize();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkForAdd(index);
+        resize();
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            resize();
+            elementData[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexOutOfBoundCheck(index);
+        checkForAdd(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexOutOfBoundCheck(index);
+        checkForAdd(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexOutOfBoundCheck(index);
+        final T elementToRemove = (T) elementData[index];
+        for (int i = index; i < size - 1; i++) {
```

Comment:
It is possible to use System.arraycopy()

Reply:
not fixed. Instead of doing copy in loop use `System.arraycopy`